### PR TITLE
Added exceptional responses handling

### PR DESCRIPTION
**Pull request by MatthewZMSU**
Now we notify client that the service struggled with exceptional response statuses

### CODE REVIEW

File: helpers/middlewares.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+exports.exceptionMiddleware = async function exceptionMiddleware(err, req, res, next) {
+    res.status(500);
+    res.send({
+        'contextId': err.contextId,
+        'pageId': err.pageId,
+    });
+    next(err.error);
```
Comment (mxsnq) [inline helpers/middlewares.js:7]: There may be errors not originating from the action function: missing context/page, invalid function for custom action, etc. Such errors are consumed and not logged.

File: helpers/middlewares.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+exports.exceptionMiddleware = async function exceptionMiddleware(err, req, res, next) {
+    res.status(500);
+    res.send({
+        'contextId': err.contextId,
+        'pageId': err.pageId,
+    });
```
Comment (mxsnq) [inline helpers/middlewares.js:6]: What do you think of sending error messages, to show them in crawler's logs?

File: helpers/middlewares.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+exports.exceptionMiddleware = async function exceptionMiddleware(err, req, res, next) {
+    res.status(500);
```
Comment (mxsnq) [inline helpers/middlewares.js:2]: Also add 'scrapy-puppeteer-service-context-id' header. This makes the context reachable if there are multiple service instances behind haproxy.

File: helpers/utils.js [hunk 4]
```diff
@@ -171,6 +171,15 @@ exports.performAction = async function performAction(request, action) {
...
+        try {
+            return await action(page, request);
+        }
+        catch (err) {
+            throw {
+                contextId: page.browserContext().id,
+                pageId: page.target()._targetId,
+                error: err,
+            }
+        }
```
Comment (mxsnq) [inline helpers/utils.js:183]: (optional) I'd suggest to [define custom Error subclass](https://github.com/goldbergyoni/nodebestpractices#-22-extend-the-built-in-error-object) or to assign these properties to the original error

File: helpers/middlewares.js (hunk 2 above)
Comment (MatthewZMSU) [reply helpers/middlewares.js:6]: Well, I don't think that crawlers need error messages as they can see this in service's logs (can they?).
One more reason NOT to add this feature is that we would print log-messages with other log-messages such as RETRY. So, we would get too many logs.
Also I'm not sure if we want to process all bad responses on our own.
At least for now I add 'error_msg' to response body in Service. You may delete it if you want.

File: helpers/utils.js (hunk 4 above)
Comment (MatthewZMSU) [reply helpers/utils.js:183]: I decided to assign properties, as this does not duplicate error message

File: helpers/middlewares.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+exports.processingMiddleware = async function processingMiddleware(err, req, res, next) {
+    if (!err.contextId) {
+        err.contextId = req.query.contextId;
+        err.pageId = req.query.pageId;
+    }
```
Comment (mxsnq) [inline helpers/middlewares.js:5]: I believe you don't need extra middleware for that, so you can move this logic to exceptionMiddleware.

If you'd like to have two middlewares, please rename processingMiddleware. One can guess what exceptionMiddleware does, but what does processingMiddleware process?

File: helpers/middlewares.js (hunk 2 above)
Comment (mxsnq) [reply helpers/middlewares.js]: > I don't think that crawlers need error messages as they can see this in service's logs (can they?).

In general crawlers do not have access to service logs.

> One more reason NOT to add this feature is that we would print log-messages with other log-messages such as RETRY. So, we would get too many logs.

I believe it is ok to log them with debug level, as retry middleware does

> Also I'm not sure if we want to process all bad responses on our own.

I agree that not all errors are meaningful to crawlers and/or should be shown to users. I'd suggest to define appropriate errors for each request/action. But for now I'm ok with the current solution.

**Pull request merged**